source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -145,6 +152,39 @@ OptRef<const LocalRateLimiterImpl::LocalDescriptorImpl> LocalRateLimiterImpl::de
 
 bool LocalRateLimiterImpl::requestAllowed(
     absl::Span<const RateLimit::LocalDescriptor> request_descriptors) const {
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.http_local_ratelimit_match_all_descriptors")) {
+    // If a request is limited by a descriptor, it should not consume token from the remaining
+    // matched descriptors, so we first check the remaining tokens of matched descriptors instead of
+    // trying to consuming their tokens. If one of these tokens is 0, it just return false and will
+    // not consume any tokens of matched descriptors. Otherwise, it will try to consume 1 token from
+    // each matched descriptor.
```

Comment:
By definition this check is never going to be perfect due to race conditions. Is this complexity worth it? It seems like not a huge deal to just consume tokens from all of them? Otherwise I think we need to carefully document this.

Reply:
This pre-check should be used to address my previous comments.

> By this way, if a request is limited by a descriptor, it will still comsume token from the remaining matched descriptors.
> For example, if client A is allowed to send 3 requests per second with a descriptor x. Then we set a gloabl tokens with limit of 10 requests per second. Then fouth request of client A shouldn't comsume the token of global tokens. But in this implementation, the global tokens still be comsumed.

For a more specific example. 
If there are two descriptors A and B. Descriptor A has more strict limit than B.
At same time, if there are two clients X and Y. Requests from X can match both A and B, and requests from Y can only match descriptor B.

Requests from X will be limited first by more strict descriptor A. But if X sends new requests continuously, these limited requests will consume tokens of descriptor B continuously. Finally, the requests from Y also will be effected.

Reply:
The problem is these checks have race conditions, and are never going to be perfect. Given that, is it really worth it to do that?

Reply:
Yeah, we can never get a accurate result because the race conditions. But I am worry about that there would be a security issue if no this pre-check.

Reply:
I just get another possible approch 🤔 . We can sort this descriptors by the limit (token per second), to make the more strict descriptors to be applied first. For every request, we can try to apply all the matched descriptors. But if the request is limited by someone, return directly.

Reply:
Yeah I like the idea of the sorting approach. That feels like a better path forward? Still though I think you will have potential inconsistency no matter what you do.

Reply:
Yeah, i also think sorting can not solve this problem perfectly. For example, we have two descriptors A and B, A is limited 3 requests per second, and B is limited 20 requests per 10 seconds. Obviously B is stricter than A (token per second), as a result, if we send requests above 3 in a second, the limited requests form A will also consume tokens of B. 

So, is it necessary to solve the problem of consuming tokens from all of these descriptors? @mattklein123 @wbpcode

Reply:
Yeah, this problem is hard to be solved perfectly. But it's still meaningful to make this feature more secure. This sorting is easy to implement and almost no side effect.
So I will vote for it.

cc @mattklein123

Reply:
In general my feeling on this is:
1. There is really no perfect solution here unless you move to explicit locking of the entire data structure and away from atomics. I suppose this is possible if correctness really matters, but that would have to be an opt-in mode.
2. If we assume that there is no perfect solution I would then bias for "good enough" and keep it as simple as possible with clear documentation on the tradeoffs.

Sorting seems simple enough to me and good enough so that would be fine as long as we clearly document it.